deptofdefense/move.mil
# PPM Estimator date input

## Checklist

I have…

- [x] run the application locally (`bin/rails server`) in a modern browser as well as in IE 11 and verified that my changes behave as expected.
- [x] Attached screenshots showing the relevant pages both before and after my changes.

## Summary of Changes

According to the USWDS, date input should be done with three separate input fields, instead of `<input type="date">` which is not implemented in IE11 or macOS Safari, or with a JavaScript datepicker, which may not be 508 compliant, or work with form validation, etc.

This changeset therefore replaces the date input on the PPM Estimator page with three fields, and performs some extra client-side date validation, checking that the "day" field is valid for the supplied month and year (leap year aware!)

Also, this changeset scrolls the page so that the results or error is in view upon clicking submit.

## Testing

To verify the changes proposed in this pull request…

1. clone this repo,
1. git checkout `ppm-estimator-date-picker`,
1. set up development dependencies according to [CONTRIBUTING.md](https://github.com/deptofdefense/move.mil/blob/master/CONTRIBUTING.md),
1. run `bin/rails server`, and
1. load up [http://localhost:3000](http://localhost:3000) in the Web browser of your choice.

## Screenshots

<details><summary>Desktop Chrome - before</summary>

<img width="537" alt="screen shot 2018-01-05 at 2 28 34 pm" src="https://user-images.githubusercontent.com/26554826/34624875-c2795046-f224-11e7-840f-91b146ccc1b4.png">

</details>

<details><summary>Desktop Chrome - after</summary>

<img width="303" alt="screen shot 2018-01-05 at 2 37 54 pm" src="https://user-images.githubusercontent.com/26554826/34625266-0cc387ce-f226-11e7-86a8-88152026cebc.png">

</details>

<details><summary>Desktop IE11 - before</summary>

![virtualbox_ie11 - win7_05_01_2018_14_37_12](https://user-images.githubusercontent.com/26554826/34625241-f40cc326-f225-11e7-86d6-6665483c568b.png)

</details>

<details><summary>Desktop IE11 - after</summary>

![virtualbox_ie11 - win7_05_01_2018_14_38_46](https://user-images.githubusercontent.com/26554826/34625294-2acae370-f226-11e7-83a3-803a6d1f624a.png)

</details>

<details><summary>Desktop Chrome - date validation error</summary>

<img width="301" alt="screen shot 2018-01-05 at 2 39 20 pm" src="https://user-images.githubusercontent.com/26554826/34625308-3de17640-f226-11e7-8203-24bcdfac8123.png">

</details>

<details><summary>iOS Safari - before</summary>

![img_25bed6bce88a-1](https://user-images.githubusercontent.com/26554826/34625498-dea46498-f226-11e7-94e7-69ce8aa42b44.jpeg)

</details>

<details><summary>iOS Safari - after</summary>

![img_07f5cbde8584-1](https://user-images.githubusercontent.com/26554826/34625557-27ca26f8-f227-11e7-9cb4-8f5c3cae2bd8.jpeg)

</details>
